fix issue with illegal reflection

pendula95 opened the pull request:
Starting from JDK 9 illegal reflective is not allowed by default. This is the reason why any java extension to syslog would fail as illegal reflection was used to load classes on class path. It was possible to use legal reflection to solve this but I try to avoid reflection when ever is possible.
Instead load classes by extending current classLoader and set it to Thread.
Fixes #1943

kira-syslogng commented:
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

kira-syslogng commented:
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

kira-syslogng commented:
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

alltilla commented:
@kira-syslogng ok to test

kira-syslogng commented:
Build FAILURE

pendula95 commented:
I am not sure why this is failing as I have built this locally and had no issue.
Even the error message does not have any sense in relation to the fix I have pushed.

alltilla commented:
Hi @pendula95,

Thank you for your PR!

 * Commits check fails for the following reason:
```
[FAIL] 9fa9969bc fix issue with illegal reflection
       * This commit subject does not match the project conventions.
       * This commit does not have a Signed-off-by line.

Please modify the commit messages to conform to this pattern:
---------------------------------- 8< ----------------------------------
modulename: short description of the commit

optional longer description

Signed-off-by: Name <email>
---------------------------------- 8< ----------------------------------

To update your commits, run `git rebase -i origin/master`, and replace
'pick' by 'reword' on each commit you wan to reword.

The "Signed-off-by" lines can be added automatically by git(1) using the
"--signoff" option when committing.  To add signed-off-by information to all
commits in your branch, run `git rebase --signoff origin/master`.
```

 * Kira failed because of a Jenkins problem, I will restart it.

Attila

alltilla commented:
@kira-syslogng: retest this please

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

gaborznagy commented:
@pendula95 Thanks for the PR again!

It seems our CI environment has some issues, that's the reason of the kira-syslogng failures.
We'll give an update once we found something!

lbudai commented:
@pendula95 : which java version are you using?
I think you should also replace `javah` with `javac -h`. (hint: #2047 )

lbudai commented:
related PR : #3433

pendula95 commented:
When testing I tried running this against java 11 and 8. Ofc there were other issue, the one you linked, https://github.com/syslog-ng/syslog-ng/issues/3053
After solving everything last issue was this reflection.
But how is the issue you linked relate to this PR?

pendula95 commented:
Issue that this PR targets to fix is problem with illegal reflection when ran against java9+ (haven't tested on all releases but did on 8 and 11)
In runtime (assuming you were able to compile the project with java 11 and ommited all issues with javah) you will get warning and then an error like this.
```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.syslog_ng.SyslogNgClassLoader (file:/home/alex/syslog-ng/lib/syslog-ng/java-modules/syslog-ng-core.jar) to method java.net.URLClassLoader.addURL(java.net.URL)
WARNING: Please consider reporting this to the maintainers of org.syslog_ng.ExampleClass
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[2020-09-21T18:52:23.009919] Error while expanding path list: object is not an instance of declaring class;
java.lang.IllegalArgumentException: object is not an instance of declaring class
.
.
.
[2020-09-21T18:52:23.010667] Can't find class; class_name='com.ipification.syslog_ng.ExampleClass'
```
Current syslog code tries to add external jar(s) using illegal reflection which is disabled starting from java9. I have updated the method for loading external resources so that it does not use reflection.

gaborznagy commented:
@kira-syslogng retest this please;

lbudai commented:
> When testing I tried running this against java 11 and 8. Ofc there were other issue, the one you linked, #3053
> After solving everything last issue was this reflection.
> But how is the issue you linked relate to this PR?

With legacy `javah` the code won't build - this is why it is related to this issue.
How did you build the code? I guess you had to do some modifications on top of our automake/cmake build files.

pendula95 commented:
> > When testing I tried running this against java 11 and 8. Ofc there were other issue, the one you linked, #3053
> > After solving everything last issue was this reflection.
> > But how is the issue you linked relate to this PR?
> 
> With legacy `javah` the code won't build - this is why it is related to this issue.
> How did you build the code? I guess you had to do some modifications on top of our automake/cmake build files.

I we made following changes to Makefile.am
```
$ git diff modules/java/Makefile.am
diff --git a/modules/java/Makefile.am b/modules/java/Makefile.am
index e612b8c69..b86297051 100644
--- a/modules/java/Makefile.am
+++ b/modules/java/Makefile.am
@@ -44,7 +44,7 @@ $(SYSLOG_NG_CORE_LIBS_DIR) $(SYSLOG_NG_CORE_CLASSES_DIR) $(SYSLOG_NG_CORE_PACKAG
        mkdir -p $@
 $(JAVA_HEADER_FILES): $(JAVA_BUILDDIR)/org_syslog_ng_%.h: $(SYSLOG_NG_CORE_PACKAGE_DIR)/%.class
-       $(AM_V_GEN)$(JAVAH) -classpath  $(SYSLOG_NG_CORE_CLASSES_DIR) -d $(JAVA_BUILDDIR) org.syslog_ng.$*
+       $(AM_V_GEN)$(JAVAC) -h $(JAVA_BUILDDIR) $(SYSLOG_NG_CORE_SRC_DIR)/*.java
 all-local: java-binaries java-headers
```

kira-syslogng commented:
Build SUCCESS

lbudai commented:
> > > When testing I tried running this against java 11 and 8. Ofc there were other issue, the one you linked, #3053
> > > After solving everything last issue was this reflection.
> > > But how is the issue you linked relate to this PR?
> > 
> > 
> > With legacy `javah` the code won't build - this is why it is related to this issue.
> > How did you build the code? I guess you had to do some modifications on top of our automake/cmake build files.
> 
> We made following changes to Makefile.am
> 
> ```
> $ git diff modules/java/Makefile.am
> diff --git a/modules/java/Makefile.am b/modules/java/Makefile.am
> index e612b8c69..b86297051 100644
> --- a/modules/java/Makefile.am
> +++ b/modules/java/Makefile.am
> @@ -44,7 +44,7 @@ $(SYSLOG_NG_CORE_LIBS_DIR) $(SYSLOG_NG_CORE_CLASSES_DIR) $(SYSLOG_NG_CORE_PACKAG
>         mkdir -p $@
>  $(JAVA_HEADER_FILES): $(JAVA_BUILDDIR)/org_syslog_ng_%.h: $(SYSLOG_NG_CORE_PACKAGE_DIR)/%.class
> -       $(AM_V_GEN)$(JAVAH) -classpath  $(SYSLOG_NG_CORE_CLASSES_DIR) -d $(JAVA_BUILDDIR) org.syslog_ng.$*
> +       $(AM_V_GEN)$(JAVAC) -h $(JAVA_BUILDDIR) $(SYSLOG_NG_CORE_SRC_DIR)/*.java
>  all-local: java-binaries java-headers
> ```

The problem is that we still support `javah` - this is why we need to add a mechanism that detects if we have `javah`, or we have a new jdk with `javac -h`. And we support not only automake but also cmake.
As the tests are passed we can merge this as-is and then we should merge #3433 .

Thanks for the PR!

lbudai commented:
@pendula95 : as it came out during our internal tests, it breaks older java versions...
So I  added your original commit to #3433 and fixed the issue.
After internal tests are passed, we will merge your patch as part of #3433.

pendula95 commented:
> @pendula95 : as it came out during our internal tests, it breaks older java versions...
> So I added your original commit to #3433 and fixed the issue.
> After internal tests are passed, we will merge your patch as part of #3433.

For which version of java did this break so I can test also?

lbudai commented:
> > @pendula95 : as it came out during our internal tests, it breaks older java versions...
> > So I added your original commit to #3433 and fixed the issue.
> > After internal tests are passed, we will merge your patch as part of #3433.
> 
> For which version of java did this break so I can test also?

java 8

pendula95 commented:
This solution was tested with both java 8 and java 11. And it worked with both version. Can you give me output of the error you get.

lbudai commented:
> This solution was tested with both java 8 and java 11. And it worked with both version. Can you give me output of the error you get.

We are testing with hdfs, elasticsearch, kafka destination drivers.
@gaborznagy : could you provide more details on the failing testcases, or the hdfs/elasticsearch/kafka versions we have?

gaborznagy commented:
@pendula95 
I'm on Ubuntu Focal and using openjdk-8 and the only thing that I see is that the `ElasticSearchDestination::init()` call fails:
```
[2020-10-02T10:37:56.349753] Using elasticsearch2() destination is deprecated please use elasticsearch-http() instead. The java-based elasticsearch2() works as before but it may be removed in the future;
[2020-10-02T10:37:56.350528] Error initializing message pipeline; plugin_name='java', location='#buffer:8:3'
```
(this is the same with our internal tests)

I've tried to debug it but didn't found much, but it's interesting that no exception occurred.
I've tried log4j but not even log4j initialized correctly (yes, I can see the same error messages with master branch too, but there I can see log4j initializes in the end).
Console log of build pendula95:fix-class-loading-java-11 with log4j enabled:
```
[2020-10-02T10:58:11.285476] WARNING: Your configuration file uses an obsoleted keyword, please update your configuration; keyword='option', change='The option() argument has been obsoleted in favour of options()', location='/home/gabornagy/install/syslog-ng/etc/syslog-ng.conf:22:3'
[2020-10-02T10:58:11.287515] Accepting connections; addr='AF_INET(127.0.0.1:1234)'
log4j: Trying to find [/home/gabornagy/install/ose/etc/log4j.properties] using context classloader java.net.FactoryURLClassLoader@6d6f6e28.
log4j: Trying to find [/home/gabornagy/install/ose/etc/log4j.properties] using java.net.FactoryURLClassLoader@6d6f6e28 class loader.
log4j: Trying to find [/home/gabornagy/install/ose/etc/log4j.properties] using ClassLoader.getSystemResource().
log4j: Could not find resource: [/home/gabornagy/install/ose/etc/log4j.properties].
[2020-10-02T10:58:11.404282] Using elasticsearch2() destination is deprecated please use elasticsearch-http() instead. The java-based elasticsearch2() works as before but it may be removed in the future;
[2020-10-02T10:58:11.405022] Error initializing message pipeline; plugin_name='java', location='#buffer:8:3'
```

The file exists, and my build from master with the very same config works.
(Again, log4j has some issues finding the log4j.properties file just as like on this branch, but later we can see that the file indeed parsed)
```
$ ls /home/gabornagy/install/ose/etc/log4j.properties
/home/gabornagy/install/ose/etc/log4j.properties

$ ~/install/ose-master/sbin/syslog-ng -Fe
[2020-10-02T11:04:53.523835] WARNING: Your configuration file uses an obsoleted keyword, please update your configuration; keyword='option', change='The option() argument has been obsoleted in favour of options()', location='/home/gabornagy/install/ose-master/etc/syslog-ng.conf:22:3'
[2020-10-02T11:04:53.526137] Accepting connections; addr='AF_INET(127.0.0.1:1234)'
log4j: Trying to find [/home/gabornagy/install/ose/etc/log4j.properties] using context classloader sun.misc.Launcher$AppClassLoader@764c12b6.
log4j: Trying to find [/home/gabornagy/install/ose/etc/log4j.properties] using sun.misc.Launcher$AppClassLoader@764c12b6 class loader.
log4j: Trying to find [/home/gabornagy/install/ose/etc/log4j.properties] using ClassLoader.getSystemResource().
log4j: Could not find resource: [/home/gabornagy/install/ose/etc/log4j.properties].
[2020-10-02T11:04:53.641483] Using elasticsearch2() destination is deprecated please use elasticsearch-http() instead. The java-based elasticsearch2() works as before but it may be removed in the future;
log4j: Parsing for [root] with value=[INFO,console].
log4j: Level token is [INFO].
log4j: Category root set to INFO
log4j: Parsing appender named "console".
log4j: Parsing layout options for "console".
log4j: Setting property [conversionPattern] to [%d{ISO8601} %p %c{2}: %m%n].
log4j: End of parsing for "console".
log4j: Setting property [target] to [System.err].
log4j: Parsed "console" options.
log4j: Finished configuring.
[main] INFO io.searchbox.client.AbstractJestClient - Setting server pool to a list of 1 servers: [http://192.168.33.10:9200]
[main] INFO io.searchbox.client.JestClientFactory - Using multi thread/connection supporting pooling connection manager
[main] INFO io.searchbox.client.JestClientFactory - Using default GSON instance
[main] INFO io.searchbox.client.JestClientFactory - Node Discovery disabled...
[main] INFO io.searchbox.client.JestClientFactory - Idle connection reaping disabled...
2020-10-02 11:04:59,525 INFO root: Failed to get cluster name from the client, use the name set in the config file: es-syslog-ng
[2020-10-02T11:04:59.529298] syslog-ng starting up; version='3.29.1.107.g1040f91'
```

config:
```
@version: 3.29
@include "scl.conf"

options {
  jvm_options("-Dlog4j.debug -Dlog4j.configuration=/home/gabornagy/install/ose/etc/log4j.properties");
};

destination d_elastic{
  elasticsearch2 ( 
    cluster("es-syslog-ng")
    client_mode("http")
    index("test_index")
    port("9200")
    server("192.168.33.10")
    type("slng_test_type") 
    flush_limit(5) 
  );  
};

```

gaborznagy commented:
Just an update, I'm currently looking into this, but it seems that the issue is around class loading.
I've used the `-verbose` JVM option to see class loading logs,and I only see a portion of them loaded.
Seemingly only standard libraries java.io, java.util, sun.net, sun.util, org.apache.log4j and org.syslog_ng can be found.

My environment: an ubuntu-focal based docker container, using openjdk-8-jdk.

pendula95 commented:
I haven't tested this with our custom java extension that we use, I haven't tested with log4j. Will need to test and get back to you. There might be an issue with multi thread apps as they don't share the same class loader from my method but will need to check.

MrAnno commented:
Closing in favor of #4054.

Thank you again for your contribution.
It's been a while... Hopefully, the above PR (which is based on your investigation) will finally fix the issue.
